src/nvim/window.c

```diff
@@ -1503,6 +1514,43 @@ bool win_valid_any_tab(win_T *win) FUNC_ATTR_PURE FUNC_ATTR_WARN_UNUSED_RESULT
   return false;
 }
 
+/// Creates new "winref" from "win"
+///
+/// @param win    The window to reference.
+winref_T winref_from(win_T *win)
+{
+  winref_T winref;
+  winref.wr_win = win;
+  winref.wr_fnum = win == NULL ? 0 : win->handle;
+  winref.wr_win_free_count = win_free_count;
+  return winref;
+}
+
+/// Store "win" in "winref" and set the free count.
+///
+/// @param winref Reference to be used for the window.
+/// @param win    The window to reference.
+void winref_set(winref_T *winref, win_T *win)
+{
+  winref->wr_win = win;
+  winref->wr_fnum = win == NULL ? 0 : win->handle;
+  winref->wr_win_free_count = win_free_count;
+}
+
+/// Return true if "winref->wr_win" points to the same window as when
+/// set_winref() was called and it is a valid window.
+/// Only goes through the window list if win_free_count changed.
+/// Also checks if b_fnum is still the same, a :bwipe followed by :new might get
+/// the same allocated memory, but it's a different window.
+///
+/// @param winref Buffer reference to check for.
+bool winref_valid(winref_T *winref)
+{
+  return winref->wr_win_free_count == win_free_count
+    ? true
+    : win_valid(winref->wr_win) && winref->wr_fnum == winref->wr_win->handle;
```

Comment:
this can be replaced by `handle_get_window(winref->wr_fnum) == winref->wr_win` (single hash lookup instead of linked list traversal)

Reply:
even more radically: skip over the addition of `winref_T` entirely and just store the `Window` handle. use `handle_get_window` and check whether it is NULL or not at the site a reference would be used.

Reply:
I get your point, however I like having `winref_T` for two reasons. One, it makes the design of the codebase orthogonal for buffers & windows. Second, it feels that it indicates more clearly that the window may be invalid. Therefore, I would like to keep `winref_T`, but with the handle thing to avoid the linked list. What do you think?